Drop workaround added in 270248@main and fix the bug properly

cdumez opened the pull request:
#### 26ef6e92d328cb5dd1314873b52e3f5ab1b592f6
<pre>
Drop workaround added in 270248@main and fix the bug properly
<a href="https://bugs.webkit.org/show_bug.cgi?id=264279">https://bugs.webkit.org/show_bug.cgi?id=264279</a>

Reviewed by Nikolas Zimmermann and Said Abou-Hallawa.

The implementation of `isSVGHiddenContainer()` was incorrect and was returning
false for RenderSVGResourceContainer types, even though RenderSVGResourceContainer
subclasses RenderSVGHiddenContainer.

Fix the isSVGHiddenContainer() implementation and drop the workaround recently
introduced in 270248@main.

* Source/WebCore/rendering/RenderLayer.cpp:
(WebCore::RenderLayer::updateAncestorDependentState):
* Source/WebCore/rendering/RenderObject.h:
(WebCore::RenderObject::isSVGHiddenContainer const):
(WebCore::RenderObject::isSVGResourceContainer const):
* Source/WebCore/rendering/svg/RenderSVGResourceContainer.h:

Canonical link: <a href="https://commits.webkit.org/270302@main">https://commits.webkit.org/270302@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/f41893b853422ee57081d6e6eb27c32dbaf0a2f0

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/25076 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/3620 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/26337 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/27194 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/23020 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/25344 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/5322 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/1058 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/23285 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/25319 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/2655 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/21646 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/27775 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/2349 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/22582 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/28717 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/22884 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/22940 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/26529 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/2289 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/597 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/3644 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/6014 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/2735 "Built successfully") | | [⏳ 🛠 jsc-mips ](https://ews-build.webkit.org/#/builders/JSC-MIPSEL-32bits-Build-EWS "Waiting in queue, processing has not started yet") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/2632 "Built successfully") | | [⏳ 🧪 jsc-mips-tests ](https://ews-build.webkit.org/#/builders/JSC-MIPSEL-32bits-Build-EWS "Waiting in queue, processing has not started yet") | 
<!--EWS-Status-Bubble-End-->

nikolaszimmermann commented on Source/WebCore/rendering/RenderObject.h:
```diff
@@ -467,7 +467,7 @@ class RenderObject : public CachedImageClient, public CanMakeCheckedPtr {
     bool isSVGForeignObject() const { return type() == Type::SVGForeignObject; }
     virtual bool isLegacySVGResourceContainer() const { return false; }
-    virtual bool isSVGResourceContainer() const { return false; }
+    bool isSVGResourceContainer() const { return type() == Type::SVGResourceClipper; }
```
So when we add more LBSE specific resources that inherit from RenderSVGResourceContainer, we just add them as additional check here, right?

cdumez replied:
Correct. And if it becomes too crazy, feel free to go back to a virtual function.

nikolaszimmermann reviewed (approved):
Looks good and matches my expectations for isSVGHiddenContainers behavior. Thank you for coming up with a proper fix.

shallawa commented on Source/WebCore/rendering/RenderObject.h:
```diff
@@ -467,7 +467,7 @@ class RenderObject : public CachedImageClient, public CanMakeCheckedPtr {
     bool isSVGForeignObject() const { return type() == Type::SVGForeignObject; }
     virtual bool isLegacySVGResourceContainer() const { return false; }
-    virtual bool isSVGResourceContainer() const { return false; }
+    bool isSVGResourceContainer() const { return type() == Type::SVGResourceClipper; }
```
I think this is now a little bit confusing. I would suggest making this function return `isSVGResourceClipper()` instead of checking then type. I would suggest also adding a FIXME comment saying more checks for resource renderers should be added once they make inherit from `RenderSVGResourceContainer` (masker, marker, pattern, filter and gradient).

cdumez replied:
I don't see how checking for `isSVGResourceClipper()` is better than `type() == Type::SVGResourceClipper`. It would certainly not be consistent with my other checks.

With regards to the FIXME, it's not clear where I'd put it. I basically need to add an ASSERT() in each renderer constructor. I plan on following up shortly to add them.

cdumez committed:
Drop workaround added in 270248@main and fix the bug properly
https://bugs.webkit.org/show_bug.cgi?id=264279

Reviewed by Nikolas Zimmermann and Said Abou-Hallawa.

The implementation of `isSVGHiddenContainer()` was incorrect and was returning
false for RenderSVGResourceContainer types, even though RenderSVGResourceContainer
subclasses RenderSVGHiddenContainer.

Fix the isSVGHiddenContainer() implementation and drop the workaround recently
introduced in 270248@main.

* Source/WebCore/rendering/RenderLayer.cpp:
(WebCore::RenderLayer::updateAncestorDependentState):
* Source/WebCore/rendering/RenderObject.h:
(WebCore::RenderObject::isSVGHiddenContainer const):
(WebCore::RenderObject::isSVGResourceContainer const):
* Source/WebCore/rendering/svg/RenderSVGResourceContainer.h:

Canonical link: https://commits.webkit.org/270302@main
